### ISSUE TITLE
Create project specific cd_user

### ISSUE THREAD

**Issue by michaelsauter** [opendevstack/ods-project-quickstarters#289]
Instead of one cd_user, we need one per project (aka initiative).

**Comment by michaelsauter** [comment]
See ods-provisioning-app#98.

**Comment by rattermeyer** [comment]
How are you going to provide the user? Does the user specifiy the cd_user when provisioning a new initiative?

**Comment by michaelsauter** [comment]
Ideally it is generated automatically I'd say. I don't know the details how it is done right now - do you see a blocker why generating would not be possible?

**Comment by rattermeyer** [comment]
Yes. Creation of account (technical or personal) often follows "regulated" processes in companies.
They also often have a specific user directory, where they do not grant "external" applications the right to simply create accounts.
I wanted to ask, because I am not sure if you are building on some assumptions that not generally hold.
AFAIK in your installation the cd_user is defined in crowd.
In our installtion, we have the requirement that all "internal and technical" users are defined in the company Azure AD to not bypass the processes around user creation.
Not sure how you do it, but the account lifecycle is subject of some audits (e.g. ISO 27000).

**Comment by michaelsauter** [comment]
@rattermeyer Good point, and something that needs to be addressed when this issue is worked on. Given what you said, I do not have a solution, but I would argue that the goal (having a separate user) still holds.

**Comment by michaelsauter** [comment]
FYI @kiwo @ungerts we need some solution for this

**Comment by stitakis** [comment]
@michaelsauter @kiwo @ungerts @rattermeyer was considered in the past to use access key to integrate Jenkins with BitBucket (https://confluence.atlassian.com/bitbucketserver0511/using-bitbucket-server/controlling-access-to-code/using-ssh-keys-to-secure-git-operations/ssh-access-keys-for-system-use) it could replace the cd_user to some extend at least in Jenkins (and avoid revealing cd_user credentials by mistake)... somehow seems to me a better integration. What do you think?

**Comment by michaelsauter** [comment]
After discussing this with @kiwo  we realised that the project specific cd_user needs to be able to read (not write) the `opendevstack` project on BitBucket. This is because the quickstarter provisioning will happen in the Jenkins instances within each iniative cd namespace. There, the `Jenkinsfile` needs to have access to read the quickstarter repo from `opendevstack`, and it needs to push the generated repo to the iniative project in BitBucket.

**Comment by michaelsauter** [comment]
I verified that the "iniative cd user" does not need to have access to the `ods-configuration` repository, and it would be very good to prevent the user from accessing that repo to avoid leaking secrets when the user is compromised.

**Comment by tjaeschke** [comment]
I added properties to define repositories, that can be read from all specified cd users.

There are two options to define this in the `application.properties`:

```
scm.global.readablerepos.opendevstack[0]=ods-jenkins-shared-library
scm.global.readablerepos.opendevstack[1]=ods-ods-quickstarters
```
and

```
scm.global.readablerepos.opendevstack=ods-jenkins-shared-library,ods-ods-quickstarters
```
The property has the format 

`scm.global.readablerepo.<project>=<comma,separated,repos>`

or

```
scm.global.readablerepos.<project>[0]=repo1
scm.global.readablerepos.<project>[1]=repo2
```
Other repos can be added by defining another readablerepo project, e.g.:

```
scm.global.readablerepos.someproject=somerepo
```

**Comment by tjaeschke** [comment]
The assumption for a specific cd user is, that the user is member of a group, that is allowed to access Bitbucket.
